### PR TITLE
Allocate bucket dynamically

### PR DESCRIPTION
**Pull request by shlomitk1**
Related to https://github.com/IBM/the-mesh-for-data/issues/104

This PR includes:
- integration with DLF
- introducing M4DStorageAccount CRD  
- removing M4DBucket
- Allocating buckets in the required location (fetching the relevant M4DStorageAccount to find details about the endpoint and credentials)
- deleting Datasets after M4DApplication has been deleted or encountered errors
- checking Dataset status prior to plotter generation
- Fetching credentials of provisioned buckets from secrets
- Removing temporary buckets 

This PR does not include:
- Dataset error handling (for example, if the bucket limit has been reached)

### CODE REVIEW

File: manager/controllers/app/storage.go [hunk 1]
```diff
@@ -6,168 +6,95 @@ package app
...
-// +kubebuilder:rbac:groups=app.m4d.ibm.com,resources=m4dbuckets,verbs=get;list;watch;update;
-// +kubebuilder:rbac:groups=app.m4d.ibm.com,resources=m4dbuckets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=app.m4d.ibm.com,resources=m4dstorageaccounts,verbs=get;list;watch;update;
+// +kubebuilder:rbac:groups=com.ie.ibm.hpsys,resources=datasets,verbs=get;list;watch;create;update;patch;delete
 
-// FindAvailableBucket finds an available storage asset - in the relevant geo
-// TODO: Implement check for geo.  Needed for ingest
-func FindAvailableBucket(c client.Client, log logr.Logger, owner types.NamespacedName, id string, prefixBase string, canShare bool, geo string) *app.M4DBucket {
-	ctx := context.Background()
+// ProvisionInterface is an interface for managing Dataset resources
```
Comment (simanadler) [inline manager/controllers/app/storage.go:23]: Maybe call out in the comment that it's an interface for managing dynamically allocated dataset resources

File: manager/controllers/app/storage.go [hunk 2]
```diff
@@ -6,168 +6,95 @@ package app
...
-// HasOwner checks whether the given owner owns the resource
-func HasOwner(resource *app.M4DBucket, owner types.NamespacedName) bool {
-	ownerID := CreateOwnerID(owner)
-	for _, val := range resource.Status.Owners {
-		if val == ownerID {
-			return true
-		}
-	}
-	return false
+type ProvisionFake struct {
```
Comment (simanadler) [inline manager/controllers/app/storage.go:44]: Why is this called fake?

File: manager/controllers/app/storage.go [hunk 3]
```diff
@@ -6,168 +6,95 @@ package app
...
-	if HasOwner(resource, owner) && len(resource.Status.Owners) == 1 {
-		return true
-	}
-	// sharing available for the given data set
-	if canShare && GetPrefix(resource, id) != "" {
-		return true
-	}
-	return false
+func generateDatasetName(owner types.NamespacedName, id string) string {
+	name := owner.Name + "-" + owner.Namespace + "-" + utils.Hash(id, 20)
```
Comment (simanadler) [inline manager/controllers/app/storage.go:98]: Should perhaps the source data set name be included in the name?

File: manager/apis/app/v1alpha1/m4dstorageaccount_types.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	// +required
+	// Secret including credentials for COS bucket
+	SecretRef string `json:"secretRef"`
+	// +required
+	// Endpoint
+	Endpoint string `json:"endpoint"`
+	// +required
+	// Region
+	Region string `json:"region"`
+}
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/m4dstorageaccount_types.go:21]: Not here. This is a general storage account where buckets are created. Some may be persistent, other may not. According the the requirements to be cataloged, we know whether the bucket is persistent or not.

File: manager/controllers/app/m4dapplication_controller.go [hunk 5]
```diff
@@ -252,12 +245,15 @@ func (r *M4DApplicationReconciler) reconcile(applicationContext *app.M4DApplicat
...
+		}
+	}
 	// generate blueprint specifications (per cluster)
 	blueprintPerClusterMap := r.GenerateBlueprints(instances, applicationContext)
-	resourceRef, err := r.ResourceInterface.CreateResourceReference(applicationContext.Name, applicationContext.Namespace)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+	resourceRef := r.ResourceInterface.CreateResourceReference(applicationContext.Name, applicationContext.Namespace)
 	ownerRef := &app.ResourceReference{Name: applicationContext.Name, Namespace: applicationContext.Namespace}
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:257]: CreateResourceReference just provides a name and a namespace of the managed resource (plotter). 
Once (when blueprints could be generated in other namespaces) this function could also create a new namespace, hence a possible error. We don't create any resources in this function, so no error should be returned.

File: manager/controllers/app/m4dapplication_controller.go [hunk 6]
```diff
@@ -252,12 +245,15 @@ func (r *M4DApplicationReconciler) reconcile(applicationContext *app.M4DApplicat
...
+			return ctrl.Result{}, err
+		}
+	}
 	// generate blueprint specifications (per cluster)
 	blueprintPerClusterMap := r.GenerateBlueprints(instances, applicationContext)
-	resourceRef, err := r.ResourceInterface.CreateResourceReference(applicationContext.Name, applicationContext.Namespace)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+	resourceRef := r.ResourceInterface.CreateResourceReference(applicationContext.Name, applicationContext.Namespace)
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:256]: The signature has changed since no error can occur in this function.

File: manager/controllers/app/storage.go (hunk 1 above)
Comment (shlomitk1) [reply manager/controllers/app/storage.go:23]: OK.

File: manager/controllers/app/storage.go (hunk 2 above)
Comment (shlomitk1) [reply manager/controllers/app/storage.go:44]: It is used for testing, not a real provisioning. I can change the name...

File: manager/controllers/app/storage.go (hunk 3 above)
Comment (shlomitk1) [reply manager/controllers/app/storage.go:98]: OK.

File: manager/apis/app/v1alpha1/m4dapplication_types.go [hunk 7]
```diff
@@ -169,6 +169,10 @@ type M4DApplicationStatus struct {
 	// Generated resource identifier
 	// +optional
 	Generated *ResourceReference `json:"generated,omitempty"`
+
+	// Provisioned storage (Dataset resources)
+	// +optional
```
Comment (simanadler) [inline manager/apis/app/v1alpha1/m4dapplication_types.go:174]: A better description of what this is would be helpful.  When is it used? It's mapping what to what?

**Review by simanadler — COMMENTED** [review]
After minor comment is addressed it's ready for merge from my perspective

File: manager/apis/app/v1alpha1/m4dapplication_types.go (hunk 7 above)
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/m4dapplication_types.go:176]: Added.

File: manager/controllers/app/m4dapplication_controller.go [hunk 8]
```diff
@@ -120,20 +123,38 @@ func (r *M4DApplicationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 	return ctrl.Result{}, nil
 }
 
-func checkResourceStatus(applicationContext *app.M4DApplication, status app.ObservedState) {
+func (r *M4DApplicationReconciler) checkReadiness(applicationContext *app.M4DApplication, status app.ObservedState) error {
```
Comment (roee88) [inline manager/controllers/app/m4dapplication_controller.go:126]: Just to be sure: the logic here only runs once after success right?

File: manager/controllers/app/moduleinstance.go [hunk 9]
```diff
@@ -53,43 +66,84 @@ Updates to add ingest:
...
 	connection, err := serde.ToRawExtension(&pb.DataStore{
 		Type: pb.DataStore_S3,
 		Name: "S3",
 		S3: &pb.S3DataStore{
-			Bucket:    bucket.Spec.Name,
-			Endpoint:  bucket.Spec.Endpoint,
-			ObjectKey: bucket.Status.AssetPrefixPerDataset[item.Context.DataSetID],
+			Bucket:    dataset.Spec.Local["bucket"],
+			Endpoint:  dataset.Spec.Local["endpoint"],
+			ObjectKey: originalAssetName + utils.Hash(originalAssetName, 10),
```
Comment (roee88) [inline manager/controllers/app/moduleinstance.go:93]: Why is the hash needed?

File: manager/controllers/app/moduleinstance.go [hunk 10]
```diff
@@ -53,43 +66,84 @@ Updates to add ingest:
...
 
 	return &app.DataStore{
-		CredentialLocation: utils.GetFullCredentialsPath(bucket.Spec.VaultPath),
+		CredentialLocation: utils.GetFullCredentialsPath(utils.GetDatasetVaultPath(bucketName)),
 		Connection:         *connection,
 		Format:             string(destinationInterface.DataFormat),
 	}, nil
 }
 
+func (m *ModuleManager) RegisterSecretInVault(id string, secretRef types.NamespacedName) error {
```
Comment (roee88) [inline manager/controllers/app/moduleinstance.go:107]: We will probably need this one for other uses, can you generalize it and move out of module manager? (e.g., `pkg/vault`)

(can be in separate PR)

**Review by roee88 — APPROVED** [review]
Overall looks good (at least what I could understand ;)).

Except for the comments, I think that it would be good to move storage related code like the ProvisionInterface out of the app controller and into `pkg/storage` as decoupled code that is used by the app controller. This will allow others to contribute to this part of the code as it evolves without messing with the app controller too much.

**Comment by roee88** [comment]
One more question, if DLF is jot installed in the cluster would flows that don't require it still work? This one is important I think.

File: Makefile [hunk 11]
```diff
@@ -55,11 +55,14 @@ cluster-prepare:
 	$(MAKE) -C third_party/cert-manager deploy
 	$(MAKE) -C third_party/registry deploy
 	$(MAKE) -C third_party/vault deploy
+	kubectl apply -f https://raw.githubusercontent.com/IBM/dataset-lifecycle-framework/master/release-tools/manifests/dlf.yaml
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply Makefile:58]: Yes, it will work on openshift, although they have a different installation for the openshift, more security oriented, I think. I'll open a discussion in the channel.

**Comment by shlomitk1** [comment]
> 
> 
> One more question, if DLF is jot installed in the cluster would flows that don't require it still work? This one is important I think.

Yes, the "read-only" flow will work. After uninstalling DLF and applying the test m4dapplication that requires only the read module, a plotter is successfully generated.

File: manager/controllers/app/moduleinstance.go (hunk 9 above)
Comment (shlomitk1) [reply manager/controllers/app/moduleinstance.go]: Typo, meant to add hash of app name+namespace. In case two apps are using the same asset, we need to separate them (for example, they can be transformed in a different way).

File: manager/controllers/app/moduleinstance.go (hunk 10 above)
Comment (shlomitk1) [reply manager/controllers/app/moduleinstance.go:106]: OK, will do in  a new PR

File: manager/controllers/app/m4dapplication_controller.go (hunk 8 above)
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:130]: checkReadiness runs on every Plotter update. The post-processing (registering data assets) runs once.

File: go.mod [hunk 12]
```diff
@@ -52,6 +53,50 @@ require (
...
+
+replace k8s.io/kubelet => k8s.io/kubelet v0.18.6
+
+replace k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.18.6
+
+replace k8s.io/metrics => k8s.io/metrics v0.18.6
+
+replace k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.18.6
+
+replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply go.mod]: Done

File: manager/apis/app/v1alpha1/m4dstorageaccount_types.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+type M4DStorageAccountSpec struct {
+	// +required
+	// Secret including credentials for COS bucket
+	SecretRef string `json:"secretRef"`
+	// +required
+	// Endpoint
+	Endpoint string `json:"endpoint"`
+	// +required
+	// Region
+	Region string `json:"region"`
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/m4dstorageaccount_types.go]: Done

**Review by simanadler — APPROVED** [review]

File: manager/apis/app/v1alpha1/m4dstorageaccount_types.go [hunk 14]
```diff
@@ -16,8 +16,9 @@ type M4DStorageAccountSpec struct {
 	// Endpoint
 	Endpoint string `json:"endpoint"`
 	// +required
-	// Region
-	Region string `json:"region"`
+	// +kubebuilder:validation:MinItems=1
+	// Regions
+	Regions []string `json:"region"`
```
Comment (roee88) [inline manager/apis/app/v1alpha1/m4dstorageaccount_types.go]: FYI I think that endpoints and regions are usually coupled but ok for now, we can change it later as needed.

**Review by roee88 — CHANGES_REQUESTED** [review]
@shlomitk1 you regenerated the charts and overwritten manual changes that @hunchback made. Please add his updates on top. Once we finish the move to Helm this won't be an issue but we don't want regressions in the meantime.

File: manager/apis/app/v1alpha1/m4dstorageaccount_types.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	// +required
+	// Secret including credentials for COS bucket
+	SecretRef string `json:"secretRef"`
+	// +required
+	// Endpoint
+	Endpoint string `json:"endpoint"`
+	// +required
+	// +kubebuilder:validation:MinItems=1
+	// Regions
+	Regions []string `json:"regions"`
```
Comment (roee88) [inline manager/apis/app/v1alpha1/m4dstorageaccount_types.go:21]: FYI I think that endpoints and regions are usually coupled but ok for now, we can change it later as needed.

Comment (shlomitk1) [reply manager/apis/app/v1alpha1/m4dstorageaccount_types.go:21]: OK, so one set of credentials for multiple <endpoint,region> pairs?

Comment (roee88) [reply manager/apis/app/v1alpha1/m4dstorageaccount_types.go:21]: I think so but I'm not sure if that's true on all/most major cloud providers so we can keep it for later PR after we verify.

File: pkg/storage/storage_provisioning.go [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+			found = true
+		} else {
+			newDatasets = append(newDatasets, d)
+		}
+	}
+	if found {
+		r.datasets = newDatasets
+		return nil
+	}
+	return errors.New("Could not delete a dataset: " + ref.Name + "\n" + message)
```
Comment (ronenkat) [inline pkg/storage/storage_provisioning.go:199]: what does "message" here means? the list of current datasets? should we reflect it to make it clear?

**Review by ronenkat — CHANGES_REQUESTED** [review]

File: charts/m4d-crd/templates/crds.yaml [hunk 17]
```diff
@@ -2003,6 +1661,72 @@ status:
...
+          description: M4DStorageAccountSpec defines the desired state of M4DStorageAccount
+          properties:
+            endpoint:
+              description: Endpoint
+              type: string
+            region:
+              description: Region
+              type: string
+            secretRef:
+              description: Secret including credentials for COS bucket
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/m4d-crd/templates/crds.yaml:1718]: We are creating Dataset CRDs with type: "COS" in the spec, so I assume it works only with COS. Especially, since they support only cos_hmac credentials.

File: manager/apis/app/v1alpha1/m4dstorageaccount_types.go [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// M4DStorageAccountStatus defines the observed state of M4DStorageAccount
+type M4DStorageAccountStatus struct {
+}
+
+// M4DStorageAccount defines a storage account used for copying data
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/m4dstorageaccount_types.go:28]: Added a comment.

File: manager/apis/app/v1alpha1/m4dstorageaccount_types.go [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// M4DStorageAccountSpec defines the desired state of M4DStorageAccount
+type M4DStorageAccountSpec struct {
+	// +required
+	// Secret including credentials for COS bucket
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/m4dstorageaccount_types.go:13]: OK.

File: manager/controllers/app/m4dapplication_controller.go [hunk 20]
```diff
@@ -171,12 +201,22 @@ func (r *M4DApplicationReconciler) reconcileFinalizers(applicationContext *app.M
...
 func (r *M4DApplicationReconciler) deleteExternalResources(applicationContext *app.M4DApplication) error {
-	// clear provisioned buckets
-	key, _ := client.ObjectKeyFromObject(applicationContext)
-	if err := r.FreeStorageAssets(key); err != nil {
-		return err
+	// clear provisioned storage
+	var deletedBuckets []string
+	var errMsgs []string
+	for _, bucketName := range applicationContext.Status.ProvisionedStorage {
+		if err := r.Provision.DeleteDataset(getBucketResourceRef(bucketName)); err != nil {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:208]: We are deleting dataset CRDs. Adding a comment saying this. Temporary buckets will be deleted as well.

File: manager/controllers/app/m4dapplication_controller.go [hunk 21]
```diff
@@ -171,12 +201,22 @@ func (r *M4DApplicationReconciler) reconcileFinalizers(applicationContext *app.M
...
+			errMsgs = append(errMsgs, err.Error())
+		} else {
+			deletedBuckets = append(deletedBuckets, bucketName)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:213]: No. Where should we log this information - in the manager log?

File: manager/controllers/app/m4dapplication_controller.go [hunk 22]
```diff
@@ -171,12 +201,22 @@ func (r *M4DApplicationReconciler) reconcileFinalizers(applicationContext *app.M
...
+	for _, bucket := range deletedBuckets {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:214]: Manager does not know whether the actual bucket has been deleted or not. It manages only Dataset resources that are running in m4d-system. If a Dataset resource is deleted successfully, while an actual bucket (non-persistent) is still alive, the manager would not know about it.

File: pkg/storage/storage_provisioning.go [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+// CreateDataset generates a Dataset resource
+func (r *ProvisionImpl) CreateDataset(ref *types.NamespacedName, bucket *ProvisionedBucket, owner *types.NamespacedName) error {
+	existing := &comv1alpha1.Dataset{}
+	if err := r.Client.Get(context.Background(), *ref, existing); err == nil {
+		if equal(bucket, &existing.Spec) {
+			// update is not required
+			return nil
+		}
+		// re-create the dataset
+		if err = r.DeleteDataset(ref); err != nil {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/storage/storage_provisioning.go:76]: Buckets are supposed to be unique per dataset + app name + app namespace. I did not check though what happens when we ask to provision a new bucket while such a bucket already exists. Currently, this PR does not handle any errors while creating a bucket, just reporting back to the user. Should be some re-attempt based on the specific error (i.e. no buckets can be created - change a storage account, or the bucket exists - think of a different name).

File: charts/m4d-crd/templates/crds.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/m4d-crd/templates/crds.yaml:1718]: any S3 object store

File: pkg/storage/storage_provisioning.go [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+			return err
+		}
+	}
+	values := map[string]string{
+		"type":             "COS",
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/storage/storage_provisioning.go:81]: It is what DLF supports. They plan to change this.

File: pkg/storage/storage_provisioning.go (hunk 16 above)
Comment (shlomitk1) [reply pkg/storage/storage_provisioning.go]: Modified.

File: pkg/storage/storage_provisioning.go [hunk 25]
```diff
@@ -0,0 +1,200 @@
+// Copyright 2020 IBM Corp.
+// SPDX-License-Identifier: Apache-2.0
+
+package storage
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/storage/storage_provisioning.go:15]: In the code.

**Comment by shlomitk1** [comment]
> 
> 
> @shlomitk1 you regenerated the charts and overwritten manual changes that @hunchback made. Please add his updates on top. Once we finish the move to Helm this won't be an issue but we don't want regressions in the meantime.

Reverted the changes, adding only necessary updates.

File: manager/controllers/app/m4dapplication_controller.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (ronenkat) [reply manager/controllers/app/m4dapplication_controller.go:214]: Lets start with the manager's log, and we will evolve it as we go along

File: Makefile [hunk 26]
```diff
@@ -55,11 +55,14 @@ cluster-prepare:
...
+
```
Comment (ronenkat) [inline Makefile:59]: how adding DLF impact the quick-start https://ibm.github.io/the-mesh-for-data/docs/setup/quickstart/ ?
if there are changes, can you add in a follow-on PR?

Comment (shlomitk1) [reply Makefile:59]: No changes are needed. The quick start uses the read module only for reading directly from the data source.
Datasets are not allocated, copy module is not applied either.

File: manager/controllers/app/m4dapplication_controller.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:214]: I think we need to log more than that - what transformations have been performed, where the modules ran, etc. I'll create a new issue for this.

[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:214]: Opened https://github.com/IBM/the-mesh-for-data/issues/323

File: Makefile (hunk 26 above)
Comment (shlomitk1) [reply Makefile:59]: Opened https://github.com/IBM/the-mesh-for-data/issues/323 for logging compliance information.

**Review by roee88 — APPROVED** [review]

**Review by ronenkat — APPROVED** [review]
Thank you.
